File: solutions/ruby/word-count/1/word_count.rb
# frozen_string_literal: true

# Count words in an input string
class Phrase
  def initialize(str)
    @str = str
  end

  def word_count
    counts = Hash.new(0)

    @str
      .scan(/\b[\w']+\b/)
      .each_entry { |el| counts[el.downcase] += 1 }

    counts
  end
end

File: solutions/elixir/word-count/1/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> String.downcase()
    |> String.replace("_", " ")
    |> then(&Regex.scan(~r/\b[\w']+\b/, &1))
    |> List.flatten
    |> Enum.frequencies()
  end
end
